modules/Bio/EnsEMBL/Compara/RunnableDB/GeneTrees/OrthoTree.pm

```diff
@@ -566,5 +590,83 @@ sub check_homology_consistency {
     $self->throw("Inconsistent homologies: $bad_key") if defined $bad_key;
 }
 
+sub _create_flatfile {
+    my $self = shift;
+
+    my $outfile = $self->param('output_flatfile');
+
+    # create directory
+    make_path(dirname($outfile));
+
+    # open file handle and print header
+    open( my $outfh, '>', $outfile ) or die "Cannot open $outfile for writing";
+    print $outfh join("\t", @{ $Bio::EnsEMBL::Compara::Homology::object_summary_headers }) . "\n";
+    $self->param('output_filehandle', $outfh);
+    return $outfh;
+}
+
+sub _hc_flatfile {
+    my $self = shift;
+
+    my $file = $self->param('output_flatfile');
+    print "running HCs on homologies:\n" if $self->debug;
+
+    # first, check the integrity of the file
+    $self->_check_file_integrity($file);
+    print "\t- file integrity ok\n" if $self->debug;
+
+    # then HC the homologies themselves
+    my $gene_member_adaptor = $self->compara_dba->get_GeneMemberAdaptor;
+
+    open( my $fh, '<', $file ) or die "Cannot open $file for reading";
+    my $header_line = <$fh>;
+    my @header_cols = split(/\s+/, $header_line);
+
+    my $c = 1;
+    my %seen_homologies;
+    while ( my $line = <$fh> ) {
+        my $row = map_row_to_header($line, \@header_cols);
+        my ( $st_id, $hom_st_id ) = ($row->{stable_id}, $row->{homology_stable_id});
+
+        # A pair of gene can only appear in 1 homology at most
+        if (my $prev_line = ($seen_homologies{"$st_id - $hom_st_id"} || $seen_homologies{"$hom_st_id - $st_id"}) ) {
+            die sprintf(
+                "Pair (%s - %s) seen more than once (lines %d and %d)",
+                ( $st_id, $hom_st_id, $prev_line, $c )
+            );
+        }
+        $seen_homologies{"$st_id - $hom_st_id"} = $c;
+        $seen_homologies{"$hom_st_id - $st_id"} = $c;
+
+        # Checks that the seq_member_id only links to canonical peptides
+        my $gene_member = $gene_member_adaptor->fetch_by_dbID($row->{gene_member_id});
+        my $hom_gene_member = $gene_member_adaptor->fetch_by_dbID($row->{homology_gene_member_id});
+        if ( $gene_member->canonical_member_id != $row->{seq_member_id} ) {
+            die sprintf(
+                "%s is not a canonical member (line %d)",
+                ($st_id, $c)
+            );
+        } elsif ( $hom_gene_member->canonical_member_id != $row->{homology_seq_member_id} ) {
+            die sprintf(
+                "%s is not a canonical member (line %d)",
+                ($hom_st_id, $c)
+            );
+        }
```

Comment:
I know this is a copy of the previous SQLHealthchecks, but I think it should now be done differently.

The first problem is performance. There will be two `fetch_by_dbID` for each homology row, so about n² in total. Independent small queries are a killer. That's why many runnables have switch to using `Utils::Preloader` and `fetch_all_by_dbID_list`. I think it's a pity to replace a n² bottleneck with another one.

The second is usefulness / redundancy. There are still some SqlHealtchecks to check we only have canonical seq_members in the database (`only_canonical` is set to 1) and the trees are only made of canonical members. It is unlikely that the `ortho_tree` will come up with some non-canonical members.

The third problem is that this Runnable doesn't really care whether the members are canonical or not. It infers homologies from a reconciled gene tree, and is expected to list homologies that match the trees, that's all.

There is still value in checking that the file is sane and member_ids are valid. I would suggest to get the list of seq_members of the gene tree, and check that all the seq_members listed in the file are part of those. This commit should work: https://github.com/Ensembl/ensembl-compara/commit/111063e4e1836544d402a48d53bbe61fe143f904

Reply:
Yeah, I'm happy to incorporate that commit. Not sure how to go about it? At this point, we kinda just need this PR to be merged so we can start production 😅